File: data-pipeline-libraries-java/data-pipeline-core-java/src/main/java/com/enrichmeai/culvert/governance/Masker.java
package com.enrichmeai.culvert.governance;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * Applies a {@link MaskingPolicy} to a single field value.
 *
 * <p>This is a pure-Java, cloud-neutral utility. It contains no GCP SDK,
 * Cloud DLP, Dataplex, or tag-resolution imports. All masking is structural —
 * based on field-name matching, not per-cell content inspection.
 *
 * <h2>Supported strategies</h2>
 * <ul>
 *   <li>{@link MaskingStrategy#NONE} — value is returned unchanged.</li>
 *   <li>{@link MaskingStrategy#FULL} — value is replaced with
 *       {@code policy.replacement()} (defaults to {@code "*"}).</li>
 *   <li>{@link MaskingStrategy#REDACTED} — same as FULL; value is replaced
 *       with {@code policy.replacement()}.</li>
 *   <li>{@link MaskingStrategy#PARTIAL} — all characters of the string
 *       representation are replaced with {@code policy.replacement()} except
 *       the last 4, which are kept. If the value has 4 or fewer characters
 *       all characters are replaced. Non-string values are converted via
 *       {@code toString()} first.</li>
 *   <li>{@link MaskingStrategy#HASH} — a deterministic SHA-256 hex digest
 *       of the string representation (prefixed by the salt). The salt from
 *       {@code policy.salt()} is used as-is; if empty, a constant
 *       process-stable default salt {@code "culvert-pii-salt"} is applied
 *       (full per-process salt generation is out of scope for this ticket).</li>
 * </ul>
 *
 * <p><strong>Null passthrough:</strong> if {@code value} is {@code null} the
 * method returns {@code null} regardless of strategy — nulls are not fabricated.
 *
 * @since Sprint 14 / T14.4
 */
public final class Masker {

    /** Default salt used when {@code MaskingPolicy.salt()} is empty. */
    private static final String DEFAULT_SALT = "culvert-pii-salt";

    private Masker() { /* utility class */ }

    /**
     * Applies {@code policy} to {@code value}.
     *
     * @param value  The field value to mask. May be {@code null} (returned
     *               unchanged).
     * @param policy The masking policy to apply. Must not be {@code null}.
     * @return The masked value, or {@code null} if {@code value} was null.
     * @throws NullPointerException if {@code policy} is null.
     */
    public static Object mask(Object value, MaskingPolicy policy) {
        if (policy == null) throw new NullPointerException("policy must not be null");
        if (value == null) return null;

        return switch (policy.strategy()) {
            case NONE -> value;
            case FULL, REDACTED -> policy.replacement();
            case PARTIAL -> maskPartial(value.toString(), policy.replacement());
            case HASH -> maskHash(value.toString(), policy.salt());
        };
    }

    // -----------------------------------------------------------------------
    // Strategy helpers
    // -----------------------------------------------------------------------

    /**
     * Masks all characters except the last 4, replacing them with
     * {@code replacement}.
     */
    private static String maskPartial(String s, String replacement) {
        char replChar = replacement.isEmpty() ? '*' : replacement.charAt(0);
        if (s.length() <= 4) {
            return String.valueOf(replChar).repeat(s.length());
        }
        String suffix = s.substring(s.length() - 4);
        return String.valueOf(replChar).repeat(s.length() - 4) + suffix;
    }

    /**
     * Returns a deterministic hex-encoded SHA-256 digest of
     * {@code (salt + value)}.
     */
    private static String maskHash(String value, String salt) {
        String effectiveSalt = (salt == null || salt.isEmpty()) ? DEFAULT_SALT : salt;
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            md.update(effectiveSalt.getBytes(StandardCharsets.UTF_8));
            byte[] digest = md.digest(value.getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder(digest.length * 2);
            for (byte b : digest) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            // SHA-256 is guaranteed present in every JRE; this branch is unreachable.
            throw new IllegalStateException("SHA-256 not available", e);
        }
    }
}

File: data-pipeline-libraries/data-pipeline-core/src/data_pipeline_core/governance_api/masker.py
"""Masker — applies a MaskingPolicy to a single field value.

Cloud-neutral utility: no GCP SDK, Cloud DLP, Dataplex, or external
imports. All masking is structural — based on field-name matching, not
per-cell content inspection.

Mirrors Java ``com.enrichmeai.culvert.governance.Masker`` (Sprint 14 / T14.4).

Supported strategies
--------------------
NONE      — value returned unchanged.
FULL      — value replaced with policy.replacement (default "*").
REDACTED  — same as FULL; value replaced with policy.replacement.
PARTIAL   — all characters replaced with policy.replacement[0] (or '*')
            except the last 4, which are kept. If the value has 4 or
            fewer characters, all characters are replaced.  Non-string
            values are converted via str() first.
HASH      — deterministic SHA-256 hex digest of (salt + value).  If
            policy.salt is empty, the constant process-stable salt
            ``"culvert-pii-salt"`` is used (mirrors Java DEFAULT_SALT).

Null passthrough
----------------
If ``value`` is ``None`` the method returns ``None`` regardless of strategy
— nulls are not fabricated.
"""

from __future__ import annotations

import hashlib
from typing import Any, Optional

from data_pipeline_core.governance_api.policies import MaskingPolicy, MaskingStrategy

# Mirrors Java ``Masker.DEFAULT_SALT`` (Masker.java:41)
_DEFAULT_SALT = "culvert-pii-salt"


def mask(value: Optional[Any], policy: MaskingPolicy) -> Optional[Any]:
    """Apply *policy* to *value* and return the masked result.

    Parameters
    ----------
    value:
        The field value to mask.  ``None`` is returned unchanged.
    policy:
        The masking policy.  Must not be ``None``.

    Returns
    -------
    The masked value, or ``None`` if *value* was ``None``.

    Raises
    ------
    TypeError
        If *policy* is ``None``.
    """
    if policy is None:
        raise TypeError("policy must not be None")
    if value is None:
        return None

    strategy = policy.strategy

    if strategy is MaskingStrategy.NONE:
        return value
    if strategy in (MaskingStrategy.FULL, MaskingStrategy.REDACTED):
        return policy.replacement
    if strategy is MaskingStrategy.PARTIAL:
        return _mask_partial(str(value), policy.replacement)
    if strategy is MaskingStrategy.HASH:
        return _mask_hash(str(value), policy.salt)

    # Unreachable for a well-formed MaskingStrategy enum, but guard anyway.
    raise ValueError(f"Unsupported MaskingStrategy: {strategy!r}")  # pragma: no cover


# ---------------------------------------------------------------------------
# Strategy helpers — mirrors Java private methods in Masker.java
# ---------------------------------------------------------------------------

def _mask_partial(s: str, replacement: str) -> str:
    """Mask all characters except the last 4.

    Mirrors Java ``Masker.maskPartial`` (Masker.java:74-81).
    """
    repl_char = replacement[0] if replacement else "*"
    if len(s) <= 4:
        return repl_char * len(s)
    suffix = s[-4:]
    return repl_char * (len(s) - 4) + suffix


def _mask_hash(value: str, salt: str) -> str:
    """Return a deterministic SHA-256 hex digest of (effective_salt + value).

    Mirrors Java ``Masker.maskHash`` (Masker.java:87-101): salt is prepended
    to value before hashing.
    """
    effective_salt = salt if salt else _DEFAULT_SALT
    digest = hashlib.sha256((effective_salt + value).encode("utf-8")).hexdigest()
    return digest
